url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/block/BlockMatrixOps.java

```diff
@@ -0,0 +1,674 @@
+import org.ejml.ops.MatrixFeatures;
+import org.ejml.ops.RandomMatrices;
+import org.ejml.simple.SimpleMatrix;
+
+import java.util.Random;
+
+
+/**
+ * Various operations on {@link BlockMatrix64F}.
+ *
+ * @author Peter Abeles
+ */
+public class BlockMatrixOps {
+
+    /**
+     * Converts a row major matrix into a row major block matrix.
+     *
+     * @param src Original DenseMatrix64F.  Not modified.
+     * @param dst Equivalent BlockMatrix64F. Modified.
+     */
+    public static void convert( DenseMatrix64F src , BlockMatrix64F dst )
+    {
+        if( src.numRows != dst.numRows || src.numCols != dst.numCols )
+            throw new IllegalArgumentException("Must be the same size.");
+
+        for( int i = 0; i < dst.numRows; i += dst.blockLength ) {
+            int blockHeight = Math.min( dst.blockLength , dst.numRows - i);
+
+            for( int j = 0; j < dst.numCols; j += dst.blockLength ) {
+                int blockWidth = Math.min( dst.blockLength , dst.numCols - j);
+
+                int indexDst = i*dst.numCols + blockHeight*j;
+                int indexSrcRow = i*dst.numCols + j;
+
+                for( int k = 0; k < blockHeight; k++ ) {
+                    System.arraycopy(src.data,indexSrcRow,dst.data,indexDst,blockWidth);
+                    indexDst += blockWidth;
+                    indexSrcRow += dst.numCols;
+                }
+            }
+        }
+    }
+
+    /**
+     * <p>
+     * Converts matrix data stored is a row major format into a block row major format in place.
+     * </p>
+     * 
+     * @param numRows number of rows in the matrix.
+     * @param numCols number of columns in the matrix.
+     * @param blockLength Block size in the converted matrix.
+     * @param data Matrix data in a row-major format. Modified.
+     * @param tmp Temporary data structure that is to be the size of a block row.
+     */
+    public static void convertRowToBlock( int numRows , int numCols , int blockLength ,
+                                          double[] data, double[] tmp )
+    {
+        int minLength = Math.min( blockLength , numRows ) * numCols;
+        if( tmp.length < minLength ) {
+            throw new IllegalArgumentException("tmp must be at least "+minLength+" long ");
+        }
+
+        for( int i = 0; i < numRows; i += blockLength ) {
+            int blockHeight = Math.min( blockLength , numRows - i);
+
+            System.arraycopy(data,i*numCols,tmp,0,blockHeight*numCols);
+
+
+            for( int j = 0; j < numCols; j += blockLength ) {
+                int blockWidth = Math.min( blockLength , numCols - j);
+
+                int indexDst = i*numCols + blockHeight*j;
+                int indexSrcRow = j;
+
+                for( int k = 0; k < blockHeight; k++ ) {
+                    System.arraycopy(tmp,indexSrcRow,data,indexDst,blockWidth);
+                    indexDst += blockWidth;
+                    indexSrcRow += numCols;
+                }
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 87:105, 162:179.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.